### ISSUE TITLE
ForRef and MapRef extension method for array

### ISSUE THREAD

**Issue by Thaina** [dotnet/corefx#5449]
It is an easy implement to let foreach for reference in array
I want it to be included in official System.Linq library

```
public delegate void RefAction<T>(ref T item);
public static T[] ForRef<T>(this T[] array,RefAction<T> action)
{
    int i = 0;
    while(i < array.Length)
    {
        action(ref array[i]);
        i++;
    }
    return array; /// return self to chain with Linq
}

public delegate void RefFunction<T,R>(ref T item,out R obj);
public static R[] MapRef<T,R>(this T[] array,RefFunction<T,R> func)
{
    int i = 0;
    var res = new R[array.Length];
    while(i < array.Length)
    {
        func(ref array[i],out res[i]);
        i++;
    }

    return res;
}
```

Also it could be great if this pattern also allowed in Dictionary and List

I want to discus first then I may create a fork and pull request by myself

**Comment by mikedn** [comment]
> It is an easy implement to let foreach for reference in array

It's easy to implement but what's the point of implementing this? Is there a use case for this that's common enough to warrant adding what's simply a shorthand for a traditional `for` loop?

**Comment by Thaina** [comment]
@mikedn Let's see
```
var rand = new Random();
var ints = new int[32];
var doubles = new double[ints.Length];
for(int i = 0;i < ints.Length; i++)
{
    ints[i] = rand.Next();
    doubles[i] = int[i] / (double)int.MaxValue;
}
```
```
var rand = new Random();
var ints = new int[32];
var doubles = ints.MapRef((ref int n,out double d) => {
    n = rand.Next();
    d = n / (double)int.MaxValue;
});
```

Linq style is better in my opinion. Not to mention about to initialize or modified array of large struct while we can iterate with ref

We don't have syntax for ref local yet so this is easy workaround

**Comment by Thaina** [comment]
Also if this would be allowed I would suggest to add this kind of iterate to List and Dictionary too

Which it would be more powerful if we could iterate list and dictionary of large struct to modified

And this kind of extension method will compatible to chain with other Linq function

**Comment by JonHanna** [comment]
Neither of these methods are pure functions, which makes them unsuitable for Linq.

You also don't show how you propose to have it work with other query providers, which makes them incomplete for Linq.

**Comment by Thaina** [comment]
@JonHanna fine

```
var rand = new Random();
var ints = new int[32];
var randomHalf = ints.ForRef((ref int n) => n = rand.Next()).Select((n) =>  n / (double)int.MaxValue).Select((d) => d < 0.5).ToArray().ForRef((b) => !b);
```

**Comment by mikedn** [comment]
> Linq style is better in my opinion

Expanding on what @JonHanna  already stated, your example is nowhere close to LINQ style. A LINQ version looks something like:

```C#
var ints = Enumerable.Range(0, 32).Select(i => rand.Next()).ToArray();
var doubles = ints.Select(i => i / (double)int.MaxValue).ToArray();
```

Nor does MapRef come close to typical `map` implementations as not only it produces a new array but it also mutates the existing one. 

I think that if you want to use mutability in your code then it is best to stick to imperative language primitives (`for` and maybe ask for a `foreach` with `ref` support) instead of trying to shoehorn mutability into functional style code. The imperative approach is actually safer and easier to read because you know exactly what is assigned to what.

**Comment by Thaina** [comment]
If you just really see my MapRef it is producing new array

And I propose ForRef for modified bulk array in place of foreach

Actually I would like foreach to support ref but they don't do it. But lambda function is capable to do it already with easily implementation

And for is not safe man

**Comment by mikedn** [comment]
> If you just really see my MapRef it is producing new array

It also allows you to modify the input array and your example does just that.

**Comment by JonHanna** [comment]
It's producing an array, but it's also mutating one.

1. It's not pure. So it's not suitable for Linq.
2. It's not applicable to all `IEnumerable<T>`. So it's not suitable for Linq.
3. You don't include a proposal as to how this would work for `IQueryable<T>` so it's not suitable for Linq.

**Comment by Thaina** [comment]
@mikedn @JonHanna I just state that it could modified and produce new at the same time. We don't need to always done that. I just make it 
And many times we may not want to produce new array, just modified current, that why I want to have this kind as optional for that purpose

If you want to produce new one without modified the old then current Select is well suit

And does current foreach work with all functionality of linq? It's not. It can take

As I said. My propose is for working with large struct when you want to modified by reference instead of select new one. And even current Select and foreach can modified object anyway if it not struct. Why you don't have any problem with that?

It just allow you don't mean you need to do it

**Comment by JonHanna** [comment]
But why would you want this in the core linq assemblies.

There's a set of assemblies that provide pure functions on the entire range of IEnumerable<T> and IQueryable<T>.

There are two methods that provide non-pure operations on a particular types that implement IEnumerable<T>, but not all, and on no IQueryable<T>.

Why would the latter become part of the former? There's no match here.

**Comment by mikedn** [comment]
> And even current Select and foreach can modified object anyway if it not struct. Why you don't have any problem with that?

Who says we don't have any problem with that? Modifying an object that comes from `foreach` is perfectly normal but many people would object to modifying objects passed to a `Select` selector. Just because something can be done doesn't mean that it should be done.

> My propose is for working with large struct when you want to modified by reference instead of select new one. 

If I'd be concerned with performance then it's likely that I wouldn't be using such an extension method anyway due to the cost of the delegate.

> It just allow you don't mean you need to do it.

A similar argument can be used to outright reject this suggestion. The language allows you to create extension methods but that doesn't mean that the framework needs to provide every potentially useful extension.

This suggestion has 2 distinct problems. One problem is that it is associated to LINQ when in fact it has nothing in common with LINQ. The other is that no serious arguments have been presented for it to stand on its own.

**Comment by JonHanna** [comment]
(@mikedn one can also mutate reference-type elements within a `Select`, but while that's against best-practice there's simply no way to prevent that, and if anyone came up with an ingenious way to block it the backwards compatibility ship would have long ago sailed.)

**Comment by Thaina** [comment]
@JonHanna  Why it need to be match just to become part of it?

It useful in general. Some case should use it. That's all. Don't need to compare to existing one while it work differently

If you talk about functionality, keyword like `from` `let` `where` is more useless. We could just collection.Select() / dictionary.Where() and so on. Why we have it as keyword which is really more important than just an extension method of corelib

Actually your logic is invalid. To let things become part of something is not because it match the thing already exist. But because it could do thing difference than existing one and fit in the empty place it left. Which I mean foreach by reference

**Comment by mikedn** [comment]
( @JonHanna Yes, the ship has sailed but if .NET/C# had something like C++'s `const` then I think there's a good chance that LINQ would have used it to prevent mutation inside queries )

**Comment by Thaina** [comment]
@mikedn @JonHanna Also I was make it pass by ref because it is efficient to pass big struct

Currently now at roslyn we talk about immutable parameter syntax (yes, a C++ const). If that done we could also used it for this to constraint my MapRef to be immutable as you wish
But while Linq also have the same hole as my proposal then it not a valid argument

Then again to have linq modified item in collection is not best practice. But it also just a best practice

I think the very best practice is to drop foreach syntax and always use functional style to do things. mutate or immutate should just use the same syntax. And change function name if you want to change it functionality

That's why I don't naming my function as Select but as ForRef and MapRef. So use mine when you want to mutable and use Select when you want it immutable

Or you think it has more appropriate namespace?

**Comment by Thaina** [comment]
Actually I want these kind of function implement by IEnumerable
I mean to add a function

```
public delegate void RefAction<T>(ref T item); 
public interface IEnumerable<T>
{
    T Current { get; }
    bool MoveNext { get; }
    void GetCurrent(RefAction action);
}
```

So my proposal will work with all IEnumerable. We could even implement foreach by ref

But this would break everything we currently have so I can't suggest to do it now

**Comment by mikedn** [comment]
> I think the very best practice is to drop foreach syntax and always use functional style to do things. mutate or immutate should just use the same syntax. And change function name if you want to change it functionality

Do you really think it helps your suggestion to drag impractical ideas the discussion? You cannot remove `foreach` from the language because that would break all the C# code written to date. Nor it makes sense to do so, no matter how hard it tries C# is not and will never be a functional language.

> Also I was make it pass by ref because it is efficient to pass big struct

And as I said previously, if performance is a concern you wouldn't be using such extensions methods anyway.

**Comment by Thaina** [comment]
To drop something is not need to remove it man. You could see that all keyword is not used in the same popularity. There are many things was deprecated and it just been there for old code. Did you remember async function? Did you see we use `await` with `Task` now but async function is not removed?

That's what it mean to drop something

And extension method is very efficient because it is actually syntactic sugar for static function. delegate is also efficient if the compiler optimize it. And even it not, to call static method once then loop through 10000 record of table row struct is more efficient than copy it and put it back

**Comment by JonHanna** [comment]
> Why it need to be match just to become part of it?

So it's somewhere where someone looking for it would find it, for one thing. Why would someone look to Linq to find something that has nothing to do with Linq?

> I think the very best practice is to drop foreach syntax and always use functional style to do things. mutate or immutate should just use the same syntax.

One of the features of functional style is that it doesn't mutate.

> So my proposal will work with all IEnumerable. We could even implement foreach by ref

You can't pass a foreach iteration variable to a `ref` parameter. Consider that it is after all a view on a read-only `Current` property. Anything done by `ref` would be thrown away if it was allowed.

**Comment by Thaina** [comment]
@JonHanna 
> One of the features of functional style is that it doesn't mutate.

That's just the motto of functional programming. And even functional programing sometimes need to do the modification work instead of get and return for performance reason. Think about modifying matrix when you just need to change only 3 float but you need to copy whole 16 float if you not modify it

> You can't pass a foreach iteration variable to a ref parameter. Consider that it is after all a view on a read-only Current property

If you really read what I write before making argument you will know that it possible because I suggest it should have GetCurrent with ref parameter to use instead of Current

**Comment by Maxwe11** [comment]
>
```c#
var rand = new Random();
var ints = new int[32];
var doubles = ints.MapRef((ref int n,out double d) => {
    n = rand.Next();
    d = n / (double)int.MaxValue;
});
```

[Here is the great answer why usage of such design should be minimized.](http://blogs.msdn.com/b/pfxteam/archive/2012/02/03/10263921.aspx)

**Comment by VSadov** [comment]
For the most part LINQ is a library of composable query-like operators that work with sequences. 
It would not be the right place for an extension that specifically targets arrays with the purpose of in-place mutation.

**Comment by Thaina** [comment]
@VSadov Currently now there is proposal about return ref. Which it should be supported on IEnumerable too. So this feature would not be only for array but can be possible on every collection that can return ref

**Comment by Thaina** [comment]
Future but closed ?

@VSadov @karelz

**Comment by karelz** [comment]
> @Thaina Future but closed ?

I don't understand your question.
I 'fixed' the Milestone to 1.2=2.0 on 2016/12/3 to reflect it was rejected (=Closed) in 2.0-release timeframe.
Or are you trying to argue it should not be rejected?

**Comment by Thaina** [comment]
@karelz Well that confusing. Because if it is in the future and has milestone then it not fully rejected, just not do it now. So should not be closed

**Comment by karelz** [comment]
@Thaina I am confused as well by your answers, can you please explain what you mean?

This issue was fully rejected - see the comment above https://github.com/dotnet/corefx/issues/5449#issuecomment-261677771 at the time it was closed. Setting milestone to 2.0 at that moment would have been nice, but almost no one does that as it is unnecessary bureaucracy. The key thing is that we NEVER close issues just because we do not have time to fix them - see our [issue guide](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/issue-guide.md#when-we-close-issues) for details.
Does it make sense?

If you believe the issue should not have been rejected, please say so - with explanation why (addressing the closing arguments above).

**Comment by Thaina** [comment]
@karelz Well, At first I just misunderstood that, if some request still have milestone future, I will not think it is "fully rejected" instead it would just be pending. That's why if it has both milestone but closed then it confused me

This is what I see, after I comment you modified the milestone so I think you just pending it

![image](https://cloud.githubusercontent.com/assets/1042507/22635212/3cdd0296-ec64-11e6-9a8b-0bd65741bf3d.png)


But your clarification already make me understand. Thank you

**Comment by Thaina** [comment]
@karelz And as for arguing this should not be closed, I was already comment that

> @VSadov Currently now there is proposal about return ref. Which it should be supported on IEnumerable too. So this feature would not be only for array but can be possible on every collection that can return ref

So I think this feature is worth because it not just for array. Instead, we should implement collection that could return ref along with this feature

**Comment by karelz** [comment]
BTW: Note that the 'true final' milestone is on the github page on right side -- it says "2.0.0" now.
The milestone change I did was FROM Future TO 1.2.0. (the Milestone was independently renamed to 2.0.0 later on). GitHub modification record of the milestone change Future -> 1.2.0 makes it confusing to understand what was changed to what.

@Thaina I admit I don't fully understand your argument nor your original proposal. I will let @VSadov and other Linq experts to chime in on that ...
